rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+- Feature Name: `error_type`
+- Start Date: 2024-11-18
+- RFC PR: [FuelLabs/sway-rfcs#43](https://github.com/FuelLabs/sway-rfcs/pull/43)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to standardize error handling in Sway so that there is a
+deliberate and clear way to produce error messages that are easily consumable by
+language tools while keeping error strings out of on-chain artifacts.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, reverting a sway program only produces an opaque error code.
+Sometimes this code is unusual enough that with the source code of the contract
+it is possible to figure out what is going on, but a lot of the time it's a
+meaningless number that only means that some failure happened somewhere.
+
+It's not reasonable to expect people to be able to debug sway programs in these
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings don't affect the final bytecode size.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Error types
+
+Error types are special enums that allow developers to signal that each variant
+is associated with an error message.
+
+An error type implements the `Error` trait, but it is not recommended that you
+do so manually.
+
+Instead, to create a new error type, use the following syntax and annotations on
+an enum:
+
+```sway
+#[error_type]
+enum MyError {
+  #[error("error A")]
+  A: (),
+  #[error("error B")]
+  B: (u64, u8),
+}
+```
+
+This will automatically implement the `Error` marker trait for this enum
+and let the compiler know what error messages correspond to what variants of
+the enum.
+
+It will also populate the ABI specification file with error messages in the
+metadata of the enum type. Allowing users to check the meaning of an error
+without the need to store and decode error messages on chain.
+
+For instance, the following contract method may return a `MyError` variant which
+will be compact but can still be associated with an error message:
+
+```sway
+fn do_something(self) -> Result<(), MyError> {
+  // ...
+  return Err(MyError::A); // user will know that this means "error A"
+  // ...
+}
+```
+
+## `panic`
```

Comment:
>I do not understand how the name of the intrinsic, if it is prefixed or not, makes any difference when embedding the error site information?

It doesn't make a difference per se. But it breaks a rule (all intrinsics start with __) and so it behaves surprisingly. I'd like to avoid surprise.

I can see how using a special syntax element like you're suggesting could be less surprising.

As for naming, I'm not really for using revert precisely because it's so widespread and having it as a keyword would pollute the namespace. Panic is less common and it actually has a Rust-like meaning here.

Reply:
After giving it a second thought, I am also now for more general `panic` over blockchain specific `revert`. `panic`ing is the semantics of error handling, `revert`ing is an implementation detail in the case of running Sway code on blockchain. It is open for different implementations in case of potentially other runtime environments that we were discussing, while keeping the same error handling semantics of an irrecoverable error.

To the name and nature of `panic`, if it is intrinsic, it should be `__panic`. This conveys the "special handling" and lowers, or even completely removes, the element of "surprise". But it's against the overall nature of intrinsics. They are meant for low-level code, libs etc. We don't want them polluting high level code like contracts. But `__panic` should and will be a regular guest in high-level code. To have `__panic` sticking out in high-level code I definitely don't like, and would, therefore, strike this name through.

Which leads us to the proposed `panic` as either "hidden" intrinsic that looks like a regular function call, or to a new language element, `panic` expression.

Having another one special case (intrinsic without `__`) that also hides the special compiler handling I dislike more then introducing a new expression type. That's the major reason I incline (slightly?) more towards the `panic` expression.

Reply:
Yeah I agree, the fact it's special does lean towards a special expression, even as that increases complexity slightly.